src/common/utils/getLeaderBoardDataFromGithub.ts

```diff
@@ -0,0 +1,134 @@
+  mambers: {
+    name: string;
+    node_id: string;
+    projects_names: {
+      url: string;
+      name: string;
+    }[];
+    avatar_url: string;
+    score: number;
+    stats: {
+      additions: number;
+      deletions: number;
+      commits: number;
+    };
+  }[];
+  since: string;
+  until: string;
+};
+
+type GithubContributorStats = z.infer<typeof gitHubContributorStatsSchema>;
+
+export function calculateTotals(
+  weeks: GithubContributorStats[number]['weeks'],
+) {
+  return weeks.reduce(
+    (acc, week) => {
+      const stats = {
+        additions: week.a + acc.stats.additions,
+        deletions: week.d + acc.stats.deletions,
+        commits: week.c + acc.stats.commits,
+      };
+      return {
+        score:
+          acc.score +
+          (stats.additions * 3 + stats.deletions * 2 + stats.commits),
+        stats,
+      };
+    },
+    { score: 0, stats: { additions: 0, deletions: 0, commits: 0 } },
+  );
+}
+
+async function getLeaderboardDataFromGithub() {
+  const since = '2024-01-05T00:00:00Z';
+  const until = '2024-04-12T00:00:00Z';
+  const leaderboard = (
+    await Promise.allSettled(
+      [
+        { owner: 'maakaf', repo: 'maakaf-website' },
+        { owner: 'maakaf', repo: 'Baas-data-provider' },
+      ].map(async ({ owner, repo }) => {
+        const url = `https://api.github.com/repos/${owner}/${repo}/stats/contributors?order=desc&until=${until}&since=${since}`;
+        const data = await fetch(url);
+        const json = await data.json();
+        return { owner, repo, json };
+      }),
+    )
+  ).reduce((acc, data) => {
+    if (data.status === 'rejected') return acc;
+    const parsedData = gitHubContributorStatsSchema.safeParse(data.value.json);
+    if (!parsedData.success) return acc;
+    const members = parsedData.data.map((c) => {
+      const calculateScoreAndTotals = calculateTotals(c.weeks);
+      return {
+        name: c.author.login,
+        node_id: c.author.node_id,
+        projects_names: {
+          url: `${data.value.owner}/${data.value.repo}`,
+          name: data.value.repo,
+        },
+        avatar_url: c.author.avatar_url,
+        score: calculateScoreAndTotals.score,
+        stats: calculateScoreAndTotals.stats,
+      };
+    });
+
+    for (const member of members) {
+      if (!acc.has(member.node_id)) {
+        acc.set(member.node_id, {
+          ...member,
+          projects_names: [member.projects_names],
+        });
+      } else {
+        const cur = acc.get(member.node_id)!;
+        acc.set(member.node_id, {
+          ...cur,
+          score: cur.score + member.score,
+          stats: {
+            additions: cur.stats.additions + member.stats.additions,
+            deletions: cur.stats.deletions + member.stats.deletions,
+            commits: cur.stats.commits + member.stats.commits,
+          },
+          projects_names: cur.projects_names.concat(member.projects_names),
+        });
+      }
+    }
+
+    return acc;
```

Comment:
break to smaller functions - 
```suggestion
  ).map(safeParse)
  .filter(data => data.success)
  .flatMap(calculateTotals)
  .reduce(addMember)
  .
  
  ...
  
  function addMember(set, memberData) {...}
```

Reply:
I understand your concern regarding the readability of the code, and splitting a large function into smaller, more modular ones is generally considered a best practice. However, I would like to respectfully disagree with the proposed change in this particular case and provide a counterargument for keeping the function as is for now.
As you mentioned, this function is responsible for aggregating data, which is a critical operation in our application. While splitting it into smaller functions may improve readability to some extent, it could also introduce unnecessary complexity and potential performance bottlenecks due to the increased number of function calls and context switching.
Moreover, since the code is still in an early stage and the structure may change as we continue to refine the requirements, premature optimization or refactoring could lead to wasted effort and potential regressions. It is often better to focus on getting the functionality right first and then address code quality and maintainability concerns once the design has stabilized.
That being said, I acknowledge the importance of code readability and maintainability, and I am open to revisiting this decision once we have a more concrete understanding of the final architecture and requirements. At that point, we can carefully evaluate the trade-offs and make an informed decision on whether splitting the function would be beneficial or not.
In the meantime, I suggest that we prioritize implementing the necessary functionality and focus on writing clear, self-documenting code with appropriate comments and variable naming conventions. This way, even with a larger function, the code will remain relatively readable and easier to understand and maintain in the short term.
I hope you find my perspective reasonable, and I look forward to further discussions on this matter as our project progresses.